File: python/2023/01.py
from __future__ import annotations

import re

NUMBERS = {
    "zero": 0,
    "one": 1,
    "two": 2,
    "three": 3,
    "four": 4,
    "five": 5,
    "six": 6,
    "seven": 7,
    "eight": 8,
    "nine": 9,
    "0": 0,
    "1": 1,
    "2": 2,
    "3": 3,
    "4": 4,
    "5": 5,
    "6": 6,
    "7": 7,
    "8": 8,
    "9": 9,
}


def make_number(a: re.Match | None, b: re.Match | None):
    assert a, "Input data invalid"
    assert b, "Input data invalid"
    return NUMBERS[a[1]] * 10 + NUMBERS[b[1]]


def run(data: list[str]):
    for numbers in "0123456789", NUMBERS:
        regex = "|".join(map(re.escape, numbers))
        begin = re.compile(f".*?({regex})").match
        end = re.compile(f".*({regex}).*?").fullmatch

        yield sum(make_number(begin(line), end(line)) for line in data)

File: zig/src/2023/01.zig
const std = @import("std");
const utils = @import("../utils.zig");

const Direction = enum {
    Left,
    Right,
};

const number_words = [_][]const u8{
    "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
};

fn findNumber(comptime allow_words: bool, comptime dir: Direction, data: []const u8) u8 {
    var index: usize = if (dir == Direction.Right) 0 else data.len - 1;

    while (true) {
        if ('0' <= data[index] and data[index] <= '9') return data[index] - '0';

        if (allow_words) {
            for (0.., number_words) |number, word| {
                if (index + word.len > data.len) continue;
                if (std.mem.eql(u8, data[index .. index + word.len], word)) {
                    return @intCast(number);
                }
            }
        }

        if (dir == Direction.Right) {
            index += 1;
        } else {
            index -= 1;
        }
    }
}

inline fn calcNumber(comptime allow_words: bool, line: []const u8) u128 {
    return 10 * findNumber(allow_words, Direction.Right, line) + findNumber(allow_words, Direction.Left, line);
}

pub fn run(input: *utils.Input) !utils.Result {
    var result = utils.Result{};

    while (input.next()) |line| {
        result[0] += calcNumber(false, line);
        result[1] += calcNumber(true, line);
    }

    return result;
}
